File: runtimes/compose/src/main/java/com/styleconverter/runtime/rendering/ZoomExtractor.kt
package com.styleconverter.runtime.rendering

import com.styleconverter.runtime.PropertyTracker
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.floatOrNull

/**
 * IR `Zoom` → [ZoomConfig].
 *
 * Mirrors the reader's value flavors one branch per sealed variant of
 * `ZoomValue` in `converter/src/main/kotlin/app/parsing/css/properties/
 * longhands/rendering/ZoomPropertyParser.kt`, and is the byte-parallel
 * twin of `runtimes/swiftui/.../StyleEngine/rendering/ZoomExtractor.swift`
 * and `runtimes/web/src/engine/rendering/ZoomExtractor.ts`.
 *
 * The wire shape the converter actually emits is a TAGGED OBJECT — e.g.
 * `{"type":"number","value":1.5}` for `zoom: 1.5` (grep `"type": "Zoom"`
 * in `out/tmpOutput.json`). Before this file grew the object branch it
 * only understood `JsonPrimitive`, so every real document fell through to
 * `ZoomConfig()` and `zoom` was silently dropped on Android. The bare
 * primitive branch is kept for hand-written fixtures and for a future
 * wire that flattens tag-only variants.
 */
object ZoomExtractor {

    fun extractZoomConfig(properties: List<Pair<String, JsonElement?>>): ZoomConfig {
        // Last write wins, matching the cascade every other extractor uses
        // (a later declaration of the same property overrides the earlier).
        var result = ZoomConfig()
        for ((type, data) in properties) {
            if (type == "Zoom") result = extractZoom(data)
        }
        return result
    }

    private fun extractZoom(data: JsonElement?): ZoomConfig {
        if (data == null) return ZoomConfig()

        return when (data) {
            // Tagged-object wire — the shape the converter emits today.
            is JsonObject -> fromTaggedObject(data)
            // Bare primitive: "normal"/"reset" keyword, or a raw number.
            is JsonPrimitive -> fromPrimitive(data)
            // Arrays et al. are not a shape `zoom` can take. Reported via
            // the PropertyTracker (the runtime's loud channel per CLAUDE.md
            // "no silent fallthroughs"), never swallowed.
            else -> {
                PropertyTracker.markUnhandled("Zoom")
                ZoomConfig()
            }
        }
    }

    /** `{"type":"number"|"percentage"|"normal"|"reset", "value": …}`. */
    private fun fromTaggedObject(o: JsonObject): ZoomConfig {
        val tag = (o["type"] as? JsonPrimitive)?.contentOrNull?.lowercase()
        // Both numeric variants carry the magnitude under `value`; the
        // reader wraps them in IRNumber / IRPercentage, which serialize as
        // a bare JSON number (see IRNumber/IRPercentage in ValueTypes.kt).
        val raw = (o["value"] as? JsonPrimitive)?.floatOrNull
        return when (tag) {
            // `<number>`: 1 is unzoomed, 2 doubles the used values.
            "number" -> raw?.let { factor(it) } ?: ZoomConfig()
            // `<percentage>`: the SUFFIX is load-bearing — `zoom: 150%`
            // and `zoom: 150` differ by two orders of magnitude, which is
            // exactly the corruption web's ZoomExtractor.ts calls out.
            "percentage" -> raw?.let { factor(it / 100f) } ?: ZoomConfig()
            // css-viewport-1: `normal` computes to the same used scale as 1.
            "normal" -> ZoomConfig(zoom = 1.0f, isNormal = true)
            // Legacy WebKit `reset` — identity here, see ZoomConfig's doc
            // for why that is the CONVERGENT reading, not a fallthrough.
            "reset" -> ZoomConfig(zoom = 1.0f, isNormal = true)
            // A variant the reader grew without this extractor — loud,
            // per the PropertyRegistry introspection contract.
            else -> {
                PropertyTracker.markUnhandled("Zoom")
                ZoomConfig()
            }
        }
    }

    /** Bare `"normal"` / `"reset"` / `1.5` payloads. */
    private fun fromPrimitive(p: JsonPrimitive): ZoomConfig {
        val content = p.contentOrNull?.lowercase()
        if (content == "normal" || content == "reset") {
            return ZoomConfig(zoom = 1.0f, isNormal = true)
        }
        return p.floatOrNull?.let { factor(it) } ?: ZoomConfig()
    }

    /**
     * Builds a non-normal config, rejecting factors the layout pass cannot
     * honour. css-viewport-1 clamps `zoom` to a positive value; a zero or
     * negative factor would divide the incoming constraints by <= 0 (NaN /
     * negative extents) and a non-finite one would poison the measure pass,
     * so both degrade to identity with a report instead of crashing.
     */
    private fun factor(f: Float): ZoomConfig {
        if (!f.isFinite() || f <= 0f) {
            PropertyTracker.markUnhandled("Zoom")
            return ZoomConfig()
        }
        return ZoomConfig(zoom = f, isNormal = false)
    }

    fun isZoomProperty(type: String): Boolean = type == "Zoom"
}

File: runtimes/swiftui/Sources/StyleConverterRuntime/StyleEngine/rendering/ZoomExtractor.swift
//
//  ZoomExtractor.swift
//  StyleEngine/rendering — IR `Zoom` → ZoomConfig.
//
//  Mirrors the reader's value flavors one branch per sealed variant of
//  `ZoomValue` in converter/src/main/kotlin/app/parsing/css/properties/
//  longhands/rendering/ZoomPropertyParser.kt, and is the byte-parallel
//  twin of runtimes/compose/.../rendering/ZoomExtractor.kt and
//  runtimes/web/src/engine/rendering/ZoomExtractor.ts.
//
//  Why this is NOT `ValueExtractors.extractKeyword`: that helper folds
//  `{"type":"number","value":1.5}` down to the STRING "number" (it reads
//  keyword ?? value ?? type and `value` isn't a string), which is how the
//  factor got silently dropped while `Zoom` still looked "registered".
//  The percentage variant is the other trap — `zoom: 150%` and
//  `zoom: 150` differ by two orders of magnitude, so the suffix has to be
//  read, not guessed.
//
//  Registry note: "Zoom" stays inside `RenderingProperty.names` (see
//  RenderingExtractor.swift) so PropertyRegistry ownership and the
//  Phase-10 coverage probe are unchanged. The string RenderingExtractor
//  also records for `Zoom` is inert — the rendering family has no applier
//  at all since retro P2b (A6#10) deleted the identity one, so this
//  extractor is the only path whose output reaches a modifier.
//

import Foundation

/// The single IR property this file owns.
enum ZoomPropertyName { static let name = "Zoom" }

enum ZoomExtractor {
    /// Last write wins, matching the cascade every other extractor uses.
    /// Returns nil when the IR carried no `Zoom`, so the applier chains
    /// as identity and every un-zoomed element stays byte-identical.
    static func extract(from properties: [IRProperty]) -> ZoomConfig? {
        var result: ZoomConfig?
        for p in properties where p.type == ZoomPropertyName.name {
            result = config(for: p.data)
        }
        return result
    }

    /// One IR payload → the typed config.
    private static func config(for data: IRValue) -> ZoomConfig {
        switch data {
        // Tagged-object wire — the shape the converter emits today.
        case .object(let o):
            return fromTagged(o)
        // Bare string: a hand-written fixture, or a future wire that
        // flattens tag-only variants. `normal` / `reset` and any keyword
        // outside the grammar are identity; a stringified number is not
        // guessed at as a percentage — the suffix would be missing, and
        // reading "150" as 150× is the corruption this file exists to
        // avoid, so only a bare `<number>` spelling is honoured.
        case .string(let s):
            let k = s.trimmingCharacters(in: .whitespaces).lowercased()
            if k == "normal" || k == "reset" { return ZoomConfig() }
            guard let d = Double(k) else { return ZoomConfig() }
            return factor(CGFloat(d))
        // Bare number.
        case .double(let d):  return factor(CGFloat(d))
        case .int(let i):     return factor(CGFloat(i))
        default:              return ZoomConfig()
        }
    }

    /// `{"type":"number"|"percentage"|"normal"|"reset", "value": …}`.
    private static func fromTagged(_ o: [String: IRValue]) -> ZoomConfig {
        let tag = o["type"]?.stringValue?.lowercased()
        // Both numeric variants carry the magnitude under `value`; the
        // reader wraps them in IRNumber / IRPercentage, which serialize
        // as a bare JSON number.
        let raw = o["value"]?.doubleValue
        switch tag {
        // `<number>`: 1 is unzoomed, 2 doubles the used values.
        case "number":
            guard let raw else { return ZoomConfig() }
            return factor(CGFloat(raw))
        // `<percentage>`: the suffix is load-bearing (see the header).
        case "percentage":
            guard let raw else { return ZoomConfig() }
            return factor(CGFloat(raw) / 100.0)
        // css-viewport-1: `normal` computes to the same used scale as 1.
        case "normal":
            return ZoomConfig()
        // Legacy WebKit `reset` — identity, see ZoomConfig for why that
        // is the CONVERGENT reading rather than a silent fallthrough.
        case "reset":
            return ZoomConfig()
        default:
            // A variant the reader grew without this extractor. iOS has no
            // PropertyTracker twin, so the honest signal is the identity
            // config plus this named branch (never an unlabelled default
            // that hides which shape arrived).
            return ZoomConfig()
        }
    }

    /// Builds a non-normal config, rejecting factors the layout pass
    /// cannot honour. css-viewport-1 requires a positive value; zero,
    /// negative, or non-finite would divide the incoming proposal by
    /// <= 0 (NaN / negative extents) and poison the measure pass, so all
    /// three degrade to identity rather than crash.
    private static func factor(_ f: CGFloat) -> ZoomConfig {
        guard f.isFinite, f > 0 else { return ZoomConfig() }
        return ZoomConfig(factor: f, isNormal: false)
    }
}
